dali/kernels/transpose/transpose_gpu_impl.cuh

```diff
@@ -0,0 +1,248 @@
+#define DALI_KERNELS_TRANSPOSE_TRANSPOSE_GPU_IMPL_CUH_
+
+#include <cuda_runtime.h>
+#include "dali/core/tensor_view.h"
+#include "dali/core/fast_div.h"
+#include "dali/kernels/transpose/transpose_gpu_def.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+namespace kernels {
+namespace transpose_impl {
+
+template <typename T>
+struct TiledTransposeDesc {
+  T *__restrict__ out;
+  const T *__restrict__ in;
+  uint64_t in_strides[kMaxNDim];
+  // the output strides are permuted and it's this permutation that defines the whole operation
+  uint64_t out_strides[kMaxNDim];
+  fast_div<uint64_t> shape[kMaxNDim];
+
+  uint32_t tiles_y;
+  uint32_t tiles_per_block;
+  uint32_t total_tiles;
+  int ndim;
+  int lanes;
+
+  fast_div<uint32_t> tiles_x;
+  fast_div<uint32_t> tiles_per_slice;
+};
+
+namespace transpose_shared {
+  extern __shared__ uint8_t shared_tmp[];
+}  // namespace transpose_shared
+
+
+template <int ndim, typename T>
+__device__ void TransposeTiledStatic(const TiledTransposeDesc<T> &desc) {
+  T (*tmp)[kTileSize][kTileSize+1] =
+      reinterpret_cast<T (*)[kTileSize][kTileSize+1]>(transpose_shared::shared_tmp);
+
+  int lanes = desc.lanes;
+
+  unsigned start_tile = blockIdx.x * desc.tiles_per_block;
+  unsigned end_tile = min(desc.total_tiles, start_tile + desc.tiles_per_block);
+  unsigned tile_in_slice = start_tile;
+  uint64_t fused_slice = ndim > 2
+    ? div_mod(tile_in_slice, start_tile, desc.tiles_per_slice)
+    : 0;
+
+  uint64_t pos[ndim];  // NOLINT
+
+  for (int d = ndim - 3; d >= 0; d--) {
+    fused_slice = div_mod(pos[d], fused_slice, desc.shape[d]);
+  }
+
+  unsigned tile_x, tile_y;
+  tile_y = div_mod(tile_x, tile_in_slice, desc.tiles_x);
+  pos[ndim - 1] = tile_x * kTileSize;
+  pos[ndim - 2] = tile_y * kTileSize;
+
+  T *out = desc.out;
+  const T *in = desc.in;
+
+  for (uint64_t tile = start_tile; tile < end_tile; tile++) {
+    uint64_t in_ofs = 0, out_ofs = 0;
+    for (int d = 0; d < ndim - 2; d++) {
+      in_ofs  += desc.in_strides[d] * pos[d];
+      out_ofs += desc.out_strides[d] * pos[d];
+    }
+
+    int64_t in_x  = pos[ndim-1] + threadIdx.x;
+    int64_t in_y  = pos[ndim-2] + threadIdx.y;
+    int64_t out_x = pos[ndim-1] + threadIdx.y;
+    int64_t out_y = pos[ndim-2] + threadIdx.x;
+
+    in_ofs  += desc.in_strides[ndim-2]  * in_y  + desc.in_strides[ndim-1]  * in_x;
+    out_ofs += desc.out_strides[ndim-2] * out_y + desc.out_strides[ndim-1] * out_x;
+
+    __syncthreads();
+    int tile_w = min(static_cast<uint64_t>(kTileSize), desc.shape[ndim-1] - pos[ndim-1]);
+    int tile_h = min(static_cast<uint64_t>(kTileSize), desc.shape[ndim-2] - pos[ndim-2]);
+    if (threadIdx.x < tile_w) {
+      for (int ty = threadIdx.y, dy = 0; ty < tile_h; ty += blockDim.y, dy += blockDim.y) {
+        #pragma unroll(4)
+        for (int lane = 0; lane < lanes; lane++) {
+          tmp[lane][ty][threadIdx.x] = in[in_ofs + desc.in_strides[ndim-2]*dy + lane];
+        }
+      }
```

Comment:
Maybe it is better to move `__syncthreads` here so it will be done after each loop iteration.

Reply:
I'll try and see.

Reply:
Reworked, I skipped more than just syncthreads.